themes/shortbread_v1/topics/boundaries.lua

```diff
@@ -59,45 +56,61 @@ themepark:add_proc('way', function(object, data)
         return
     end
 
-    local info = rinfos[object.id]
-    if not info then
+    local admin_level = min_admin_level[object.id]
+    if not admin_level then
         return
     end
 
     local t = object.tags
+
+    -- Set disputed flag either from disputed tag on the way...
+    local disputed = (t.disputed == 'yes')
+
+    -- .. or from a parent relation with boundary=disputed
+    if min_disputed_admin_level[object.id] and min_disputed_admin_level[object.id] <= admin_level then
+        disputed = true
+    end
+
     local a = {
-        admin_level = info.admin_level,
+        admin_level = admin_level,
         maritime = (t.maritime and (t.maritime == 'yes' or t.natural == 'coastline')),
-        disputed = info.disputed or (t.disputed and t.disputed == 'yes'),
+        disputed = disputed,
         geom = object:as_linestring()
     }
-    themepark.themes.core.add_name(a, object)
+
     themepark:insert('boundaries', a, t)
 end)
 
 themepark:add_proc('select_relation_members', function(relation)
-    if valid_admin_level(get_admin_level(relation.tags)) then
+    if is_admin_boundary(relation.tags) then
         return { ways = osm2pgsql.way_member_ids(relation) }
     end
 end)
 
 themepark:add_proc('relation', function(object, data)
-    local t = object.tags
+    if is_admin_boundary(object.tags) then
+        local admin_level = tonumber(object.tags.admin_level)
 
-    local admin_level = get_admin_level(t)
+        for _, id in ipairs(osm2pgsql.way_member_ids(object)) do
+            if not min_admin_level[id] or min_admin_level[id] > admin_level then
+                min_admin_level[id] = admin_level
+            end
+        end
 
-    if not valid_admin_level(admin_level) then
         return
     end
 
-    for _, member in ipairs(object.members) do
-        if member.type == 'w' then
-            if not rinfos[member.ref] then
-                rinfos[member.ref] = { admin_level = admin_level }
-            elseif rinfos[member.ref].admin_level > admin_level then
-                rinfos[member.ref].admin_level = admin_level
+    if object.tags.boundary == 'disputed' then
+        -- Ways in relations tagged boundary=disputed are flagged as disputed
+        -- if either the relation doesn't have an admin_level tag or the
+        -- admin_level tag is <= the admin level the way got from the
+        -- boundary=administrative relation(s).
+        local admin_level = tonumber(object.tags.admin_level or 1)
```

Comment:
Still not right, because `admin_level` will now be nil when object.tags.admin_level is set but is not a number. Must be: `local admin_level = tonumber(object.tags.admin_level or 1) or 1`

Reply:
I have added an explicit function instead which makes it more explicit what's happening here.

```
-- Get numerical admin level from string, default to 1 if invalid
local function get_admin_level(value)
    if not value or not string.match(value, '^[1-9][0-9]?$') then
        return 1
    end

    return tonumber(value)
end
```